Kernel/Net/NetworkTask.cpp

```diff
@@ -606,4 +608,20 @@ void handle_tcp(const IPv4Packet& ipv4_packet, const Time& packet_timestamp)
     }
 }
 
+void retransmit_tcp_packets()
+{
+    Vector<RefPtr<TCPSocket>, 16> sockets;
+    {
+        Locker locker(TCPSocket::sockets_for_retransmit().lock(), LockMode::Shared);
+        for (auto& socket : TCPSocket::sockets_for_retransmit().resource()) {
+            // We must keep the socket alive until after we've unlocked the hash table
+            // in case retransmit_packets() realizes that it wants to close the socket.
+            sockets.append(socket);
+
+            Locker socket_locker(socket->lock());
+            socket->retransmit_packets();
+        }
+    }
```

Comment:
Would it make sense to copy out the sockets to the vector as a separate first pass?

```cpp
NonnullRefPtrVector<TCPSocket, 16> sockets;
{
    Locker locker(TCPSocket::sockets_for_retransmit().lock(), LockMode::Shared);
    for (auto& socket : TCPSocket::sockets_for_retransmit().resource())
        sockets.append(*socket);
}
for (auto& socket : sockets) {
    Locker locker(socket.lock());
    socket.retransmit_packets();
}
```

Reply:
Yup, it would certainly help with lock contention. Fixed. :)